cibyl/features/__init__.py

```diff
@@ -0,0 +1,156 @@
+"""
+#    Copyright 2022 Red Hat
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+
+import logging
+import os
+import re
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from importlib.util import module_from_spec, spec_from_file_location
+from inspect import getmembers, isabstract, isclass
+
+from cibyl.exceptions.cli import InvalidArgument
+from cibyl.exceptions.source import NoSupportedSourcesFound, SourceException
+from cibyl.sources.source import (select_source_method,
+                                  source_information_from_method)
+from cibyl.utils.colors import Colors
+
+LOG = logging.getLogger(__name__)
+
+MODULE_PATTERN = re.compile(r"([a-zA-Z][a-zA-Z_]*)\.py")
+
+
+def is_feature_class(symbol):
+    """Check whether the symbols imported from a module correspond to
+    classes defining a feature. We assume that the symbol in question
+    defines a feature if it is a concrete class and is defined inside a module
+    within a subpackage name features.
+
+    :param symbol: An imported symbol to check
+    :type symbol: object
+    :returns: Whether the symbol defines a feature
+    :rtype: bool
+    """
+    is_concrete_class = isclass(symbol) and not isabstract(symbol)
+    return is_concrete_class and "features" in symbol.__module__
+
+
+class FeaturesLoader:
+    """Load all features available, either general or product specific through
+    some plugin."""
+    # __path__ is a list
+    features_locations = __path__
```

Comment:
The main way to introduce new paths for features will be thorugh plugins, that's way it's left as a class attribute. That's said I can certainly see the value in having an option to pass paths through the constructor, I'll add it.

Reply:
I still do not like this solution. That is an static variable, which means that if you make a change to it you will be affecting all instances of the class. That is an easy error to happen and one that can be difficult to debug. I would suggest that you either go with a fully encapsulated object or you have this as a python module without classes, just functions and global variables.

The funny thing about Python, is that depending on the type the variable is shared by all instances or not:
- feature_locations = 1 - This will not
- feature_locations = [] - But this will

Reply:
Yes, I'm aware of that, I did not think that would be a problem really, since I can't imagine a reason to have two instances of the class. Another option would be to make the FeatureLoader a Singleton class, but having everything at the module level is probably simpler and would achieve the same goal.